File: ADT/queue/Circular queue.py
# Circular queue implementation in python
# A circular queue is a fixed-size queue that wraps around when it reaches the end

# Global variables to maintain queue state
max_length = 0  # Maximum size of the queue
front_pointer = 0  # Points to the front element of the queue
end_pointer = -1  # Points to the last element of the queue
noOfElements = 0  # Tracks current number of elements in the queue
queue = []  # The actual queue array

def create():
    # """Initialize the circular queue with user-specified size"""
    global max_length, front_pointer, end_pointer, queue
    max_length = int(input("Enter the maximum size of the queue you want"))
    # Initialize empty queue with placeholder values
    queue = ['' for x in range(max_length)]
    # print(f'This is the queue you defined {queue}')

def enqueue():
    # """Add an element to the rear of the circular queue"""
    global end_pointer, front_pointer, queue, noOfElements, max_length
    if noOfElements != max_length:  # Check if queue is not full
        value = input("Enter the value of want to enter \n")
        # Calculate next position and add element
        queue[end_pointer + 1] = value
        # Update end_pointer using modulo to wrap around if necessary
        end_pointer = (end_pointer + 1) % max_length
        print(f'added {value} the queue is now {queue} ')
        noOfElements = noOfElements + 1
    else:
        print("Sorry the queue is already full")

def dequeue():
    # """Remove and return the front element from the circular queue"""
    global front_pointer, end_pointer, queue, noOfElements, max_length
    if noOfElements != 0:  # Check if queue is not empty
        # Get front element and clear its position
        value = queue[front_pointer]
        queue[front_pointer] = ''
        # Update front_pointer using modulo to wrap around if necessary
        front_pointer = (front_pointer + 1) % max_length
        noOfElements -= 1
        print(f'removed {value} the queue is now {queue}')
    else:
        print("The queue is empty")

def main():
    # """Test the circular queue implementation"""
    create()    
    enqueue()
    dequeue()

if __name__ == "__main__":
    main()

File: ADT/queue/CircularQueue.java
import java.util.Arrays;
import java.util.Scanner;

public class CircularQueue {
    static int maxLength = 0;
    static int frontPointer = 0;
    static int endPointer = -1;
    static String[] queue;
    static int noOfElements = 0;
    static Scanner scanner = new Scanner(System.in);

    // defining the create method
    public static void create() {
        System.out.println("Enter the maximum size of the queue you want");
        maxLength = scanner.nextInt();
        queue = new String[maxLength];
    }

    public static void enQueue() {
        if (noOfElements != maxLength) {
            System.out.println("Enter the data you want to add in the queue \n");
            int value = 0;
            try {
                value = scanner.nextInt();
                queue[endPointer + 1] = Integer.toString(value);
                noOfElements += 1;
            } catch (Exception e) {
                System.out.println("An error occured, are you sure you entered an integer? ");
            }
            endPointer = (endPointer + 1) % maxLength;
            System.out.println("Value " + value + " is added the queue is now: " + Arrays.toString(queue));

        } else {
            System.out.println(" The  queue is now full you cannot add any more values");
        }
    }

    public static void deQueue() {
        if (noOfElements != 0) {
            String value = queue[frontPointer];
            queue[frontPointer] = " ";

            frontPointer = ((frontPointer + 1) % maxLength);
            noOfElements -= 1;
            System.out.println("Removed " + value + " the queue is now " + Arrays.toString(queue));
        } else {
            System.out.println(" The queue is empty , what are you even trying to remove lol");
        }
    }

    public static void main(String[] args) {
        create();
        enQueue();
        deQueue();

    }
}
